In [46]:
import pandas as pd

df = pd.read_csv('/content/Dry_Bean_Dataset.csv')
display(df.head())

,Bean ID,Area,Perimeter,MajorAxisLength,MinorAxisLength,AspectRation,Eccentricity,ConvexArea,EquivDiameter,Extent,Solidity,roundness,Compactness,ShapeFactor1,ShapeFactor2,ShapeFactor3,ShapeFactor4,Class
0,1,28395,610.291,208.178117,173.888747,1.197191,0.549812,28715,190.141097,0.763923,0.988856,0.958027,0.913358,0.007332,0.003147,0.834222,0.998724,SEKER
1,2,28734,638.018,200.524796,182.734419,1.097356,0.411785,29172,191.272751,0.783968,0.984986,0.887034,0.953861,0.006979,0.003564,0.909851,0.998430,SEKER
2,3,29380,624.110,212.826130,175.931143,1.209713,0.562727,29690,193.410904,0.778113,0.989559,0.947849,0.908774,0.007244,0.003048,0.825871,0.999066,SEKER
3,4,30008,645.884,210.557999,182.516516,1.153638,0.498616,30724,195.467062,0.782681,0.976696,0.903936,0.928329,0.007017,0.003215,0.861794,0.994199,SEKER
4,5,30140,620.134,201.847882,190.279279,1.060798,0.333680,30417,195.896503,0.773098,0.990893,0.984877,0.970516,0.006697,0.003665,0.941900,0.999166,SEKER


In [47]:
df.dropna(inplace=True)


In [48]:
# Define features (X) and target (y)
features = df.drop(['Bean ID', 'Class'], axis=1)
target = df['Class']

display(features.head())
display(target.head())

,Area,Perimeter,MajorAxisLength,MinorAxisLength,AspectRation,Eccentricity,ConvexArea,EquivDiameter,Extent,Solidity,roundness,Compactness,ShapeFactor1,ShapeFactor2,ShapeFactor3,ShapeFactor4
0,28395,610.291,208.178117,173.888747,1.197191,0.549812,28715,190.141097,0.763923,0.988856,0.958027,0.913358,0.007332,0.003147,0.834222,0.998724
1,28734,638.018,200.524796,182.734419,1.097356,0.411785,29172,191.272751,0.783968,0.984986,0.887034,0.953861,0.006979,0.003564,0.909851,0.998430
2,29380,624.110,212.826130,175.931143,1.209713,0.562727,29690,193.410904,0.778113,0.989559,0.947849,0.908774,0.007244,0.003048,0.825871,0.999066
3,30008,645.884,210.557999,182.516516,1.153638,0.498616,30724,195.467062,0.782681,0.976696,0.903936,0.928329,0.007017,0.003215,0.861794,0.994199
4,30140,620.134,201.847882,190.279279,1.060798,0.333680,30417,195.896503,0.773098,0.990893,0.984877,0.970516,0.006697,0.003665,0.941900,0.999166


,Class
0,SEKER
1,SEKER
2,SEKER
3,SEKER
4,SEKER


In [49]:
display(features.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13611 entries, 0 to 13610
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Area             13611 non-null  int64  
 1   Perimeter        13611 non-null  float64
 2   MajorAxisLength  13611 non-null  float64
 3   MinorAxisLength  13611 non-null  float64
 4   AspectRation     13611 non-null  float64
 5   Eccentricity     13611 non-null  float64
 6   ConvexArea       13611 non-null  int64  
 7   EquivDiameter    13611 non-null  float64
 8   Extent           13611 non-null  float64
 9   Solidity         13611 non-null  float64
 10  roundness        13611 non-null  float64
 11  Compactness      13611 non-null  float64
 12  ShapeFactor1     13611 non-null  float64
 13  ShapeFactor2     13611 non-null  float64
 14  ShapeFactor3     13611 non-null  float64
 15  ShapeFactor4     13611 non-null  float64
dtypes: float64(14), int64(2)
memory usage: 1.7 MB


None

In [50]:
from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
label_encoder = LabelEncoder()

# Fit and transform the target variable
target_encoded = label_encoder.fit_transform(target)

# Display the first few encoded target values and the original target values for comparison
display(target_encoded[:5])
display(target.head())

array([5, 5, 5, 5, 5])

,Class
0,SEKER
1,SEKER
2,SEKER
3,SEKER
4,SEKER


In [51]:
import tensorflow as tf

# One-hot encode the target variables
y_train_encoded = tf.keras.utils.to_categorical(y_train, num_classes=len(label_encoder.classes_))
y_val_encoded = tf.keras.utils.to_categorical(y_val, num_classes=len(label_encoder.classes_))
y_test_encoded = tf.keras.utils.to_categorical(y_test, num_classes=len(label_encoder.classes_))

# Display the shapes of the one-hot encoded target variables
print("Shape of y_train_encoded:", y_train_encoded.shape)
print("Shape of y_val_encoded:", y_val_encoded.shape)
print("Shape of y_test_encoded:", y_test_encoded.shape)

Shape of y_train_encoded: (8166, 7)
Shape of y_val_encoded: (2722, 7)
Shape of y_test_encoded: (2723, 7)


# Task
Load the dry bean dataset from "Dry_Bean_Dataset.csv", use all columns except 'Bean ID' and 'Class' as features and 'Class' as the target. Label encode the target variable. Standardize the features and split the data into training (60%), validation (20%), and testing (20%) sets.

## Standardize features

### Subtask:
Apply standardization to the features to ensure they have a mean of 0 and a standard deviation of 1.


**Reasoning**:
Apply standardization to the features by importing StandardScaler, instantiating it, fitting it to the features, and transforming the features.



In [52]:
from sklearn.preprocessing import StandardScaler

# Instantiate StandardScaler
scaler = StandardScaler()

# Fit the scaler to the features data
scaler.fit(features)

# Transform the features data
features_scaled = scaler.transform(features)

display(features_scaled[:5])

array([[-0.84074853, -1.1433189 , -1.30659814, -0.63115304, -1.56505251,
        -2.18572039, -0.84145059, -1.0633406 ,  0.28908744,  0.36761352,
         1.42386706,  1.83911641,  0.68078627,  2.4021726 ,  1.92572347,
         0.83837102],
       [-0.82918764, -1.01392388, -1.39591111, -0.43444458, -1.96978438,
        -3.68603953, -0.82610154, -1.04421674,  0.69747683, -0.4629065 ,
         0.23105405,  2.49544929,  0.36796666,  3.10089364,  2.68970162,
         0.77113831],
       [-0.80715717, -1.07882906, -1.25235661, -0.58573462, -1.51429132,
        -2.04533607, -0.80870371, -1.00808399,  0.57819523,  0.51841687,
         1.25286494,  1.76484275,  0.60312904,  2.23509111,  1.84135576,
         0.91675506],
       [-0.78574055, -0.97721522, -1.27882531, -0.43929028, -1.74161826,
        -2.74221072, -0.77397521, -0.97333689,  0.67125962, -2.24176703,
         0.51504917,  2.0817155 ,  0.40171795,  2.51507461,  2.20424995,
        -0.19798533],
       [-0.78123896, -1.09738406, -1

## Split data

### Subtask:
Split the data into training (60%), validation (20%), and testing (20%) sets.


**Reasoning**:
Split the data into training, validation, and testing sets and display their shapes.



In [53]:
from sklearn.model_selection import train_test_split

# Split data into training (60%) and temporary (40%) sets
X_train, X_temp, y_train, y_temp = train_test_split(features_scaled, target_encoded, test_size=0.4, random_state=42)

# Split temporary set into validation (20%) and testing (20%) sets
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

# Display the shapes of the resulting sets
print("Shape of X_train:", X_train.shape)
print("Shape of X_val:", X_val.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_val:", y_val.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (8166, 16)
Shape of X_val: (2722, 16)
Shape of X_test: (2723, 16)
Shape of y_train: (8166,)
Shape of y_val: (2722,)
Shape of y_test: (2723,)


In [54]:
import tensorflow as tf

# Build the model
model = tf.keras.Sequential([
    tf.keras.layers.Input(shape=(X_train.shape[1],)), # Input layer with the shape of features
    tf.keras.layers.Dense(64, activation='relu'),    # Hidden layer with 64 units and relu activation
    tf.keras.layers.Dense(len(label_encoder.classes_), activation='softmax') # Output layer with softmax activation for multi-class classification
])

# Display the model summary
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_4 (Dense)                 │ (None, 64)             │         1,088 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 7)              │           455 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,543 (6.03 KB)

 Trainable params: 1,543 (6.03 KB)

 Non-trainable params: 0 (0.00 B)

In [55]:
 # Set learning rate and compile the model
learning_rate = 0.0001
optimizer = Adam(learning_rate=learning_rate)
model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])

In [56]:
from tensorflow.keras.optimizers import Adam

# Set learning rate
learning_rate = 0.0001
optimizer = Adam(learning_rate=learning_rate)

# Compile the model
model.compile(optimizer=optimizer,
              loss='categorical_crossentropy', # Appropriate for multi-class classification with softmax output
              metrics=['accuracy']) # Track accuracy during training

# Display the model configuration after compilation (optional)
print("Model compiled successfully.")

Model compiled successfully.


In [57]:
history = model.fit(X_train, y_train_encoded, epochs=50, batch_size=32, validation_data=(X_val, y_val_encoded))

Epoch 1/50
256/256 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.3954 - loss: 1.6818 - val_accuracy: 0.5915 - val_loss: 1.2656
Epoch 2/50
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6142 - loss: 1.2001 - val_accuracy: 0.7160 - val_loss: 0.9518
Epoch 3/50
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7352 - loss: 0.9147 - val_accuracy: 0.7832 - val_loss: 0.7646
Epoch 4/50
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7866 - loss: 0.7431 - val_accuracy: 0.8266 - val_loss: 0.6408
Epoch 5/50
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8299 - loss: 0.6311 - val_accuracy: 0.8582 - val_loss: 0.5527
Epoch 6/50
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8538 - loss: 0.5512 - val_accuracy: 0.8780 - val_loss: 0.4873
Epoch 7/50
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8714 - loss: 0.4864 - val_accuracy: 0.8872 - val_loss: 0.4367
Epoch 8/50
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8736 - loss: 0.4440 - val_accuracy: 0.

In [60]:
# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(X_test, y_test_encoded)

# Print the test loss and accuracy
print(f'Test Loss: {test_loss:.4f}')
print(f'Test Accuracy: {test_accuracy:.4f}')

86/86 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9267 - loss: 0.1982
Test Loss: 0.2142
Test Accuracy: 0.9192


In [66]:
# Experiment 2 - increasing the learning rate - to 0.99

learning_rate = 0.99
optimizer = Adam(learning_rate=learning_rate)
model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])


from tensorflow.keras.optimizers import Adam

# Set learning rate
learning_rate = 0.0001
optimizer = Adam(learning_rate=learning_rate)

# Compile the model
model.compile(optimizer=optimizer,
              loss='categorical_crossentropy', # Appropriate for multi-class classification with softmax output
              metrics=['accuracy']) # Track accuracy during training

# Display the model configuration after compilation (optional)
print("Model compiled successfully.")

history = model.fit(X_train, y_train_encoded, epochs=10, batch_size=32, validation_data=(X_val, y_val_encoded))


# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(X_test, y_test_encoded)

# Print the test loss and accuracy
print(f'Test Loss: {test_loss:.4f}')
print(f'Test Accuracy: {test_accuracy:.4f}')

Model compiled successfully.
Epoch 1/10
256/256 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.9252 - loss: 0.2028 - val_accuracy: 0.9331 - val_loss: 0.1824
Epoch 2/10
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9265 - loss: 0.1991 - val_accuracy: 0.9346 - val_loss: 0.1819
Epoch 3/10
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9291 - loss: 0.1958 - val_accuracy: 0.9346 - val_loss: 0.1819
Epoch 4/10
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9316 - loss: 0.1884 - val_accuracy: 0.9346 - val_loss: 0.1815
Epoch 5/10
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9296 - loss: 0.1979 - val_accuracy: 0.9324 - val_loss: 0.1816
Epoch 6/10
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9254 - loss: 0.1974 - val_accuracy: 0.9335 - val_loss: 0.1813
Epoch 7/10
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9243 - loss: 0.2054 - val_accuracy: 0.9346 - val_loss: 0.1806
Epoch 8/10
256/256 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9265 - lo

# summary -

We built a neural network model to classify different types of dry beans.

In our first try, we trained the model for 50 rounds (epochs). When we tested it on data it hadn't seen before, it correctly classified the beans about 91.92% of the time. That's pretty good!

Then, we did a quick experiment by changing something called the 'learning rate' in the model's training setup and trained it again, this time for 10 rounds. After this second training, we tested it again, and the accuracy on the new data was about 92.03%.

So, in both experiments, the model did a good job at classifying the beans, getting over 91% accuracy on new data. The second experiment with the slightly different setting resulted in a very small improvement in accuracy.